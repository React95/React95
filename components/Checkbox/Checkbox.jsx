import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import check from '../../assets/pattern/check.svg';
import checkDisabled from '../../assets/pattern/check-disabled.svg';
import bgpattern from '../../assets/pattern/dropdown.png';

const checked = ({ checked, disabled }) => checked
  && `
  &:before {
    ${
  disabled
    ? `
      background-image:  url('${checkDisabled}'), url('${bgpattern}');
      background-repeat: no-repeat, repeat;
      background-position: center center;
      background-size: 7px 7px, 1.9px 1.9px;
      `
    : `background-image: url('${check}');
      background-repeat: no-repeat;
      background-position: center center;
      background-size: 7px 7px;`
}
  }`;

const disabled = ({ disabled }) => disabled
  && `
  color: #868686;
  text-shadow: 0.5px 0.5px #d2d2d2;

  &:before {
    background-color: #c0c0c0;
  }
`;

const Field = styled.input`
  display: none;
`;

const Label = styled.label`
  position: relative;
  padding-left: 22px;
  height: 15px;
  line-height: 1.5;
  display: inline-block;
  margin: 4px 0;

  &:before {
    content: '';
    position: absolute;
    left: 0;
    display: inline-block;
    height: 12px;
    width: 12px;

    border-left: 1px solid #868a8e;
    border-top: 1px solid #868a8e;

    box-shadow: inset -1px -1px 0 0 #c3c7cb, inset 1px 1px 0 0 #000000,
      0.5px 0.5px 0 0.5px #ffffff;
    background-color: white;
  }

  ${checked} ${disabled};
`;

const Checkbox = ({
  children,
  label,
  value,
  style,
  checked,
  disabled,
  onChange,
}) => (
  <Label {...style} checked={checked} disabled={disabled}>
    <Field type="checkbox" value={value} onChange={onChange} />
    {children || label}
  </Label>
);

Checkbox.propTypes = {
  value: PropTypes.string,
  label: PropTypes.string,
  children: PropTypes.string,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

Checkbox.defaultProps = {
  value: 'checked',
  label: 'Checkbox',
  children: '',
  checked: false,
  disabled: false,
  onChange: () => {},
};

export default Checkbox;
