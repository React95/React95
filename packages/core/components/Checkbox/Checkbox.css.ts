import check from './imgs/check.svg';
import checkDisabled from './imgs/check-disabled.svg';
import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';

export const icon = style({
  width: contract.space[12],
  height: contract.space[12],
  display: 'inline-block',
  position: 'absolute',
  left: '0',
  top: contract.space[2],
  borderLeftStyle: 'solid',
  borderLeftWidth: contract.space[1],
  borderLeftColor: contract.colors.borderDark,
  borderTopStyle: 'solid',
  borderTopWidth: contract.space[1],
  borderTopColor: contract.colors.borderDark,
  backgroundColor: contract.colors.inputBackground,
  backgroundRepeat: 'no-repeat',
  backgroundPosition: 'center center',
  backgroundSize: '7px 7px',
  boxShadow: contract.shadows.input,
});

export const text = style({
  padding: contract.space[1],
  userSelect: 'none',
});

export const field = style({
  width: contract.space[12],
  height: contract.space[12],
  margin: '0',
  position: 'absolute',
  top: contract.space[2],
  left: '0',
  opacity: 0,
  selectors: {},
});

export const label = style({
  display: 'inline-block',
  height: contract.space[15],
  lineHeight: 1.5,
  position: 'relative',
  marginBlock: contract.space[4],
  paddingLeft: contract.space[20],
  selectors: {
    '&:has(:disabled)': {
      color: contract.colors.borderDark,
      textShadow: `0.5px 0.5px ${contract.colors.borderLight}`,
    },
  },
});

globalStyle(`${field}:checked + ${icon}`, {
  backgroundImage: `url('${check}')`,
});

globalStyle(`${field}:focus ~ ${text}, ${field}:active ~ ${text}`, {
  borderWidth: contract.space[1],
  borderStyle: 'dotted',
  padding: '0',
});

globalStyle(`${field}:checked:disabled + ${icon}`, {
  backgroundImage: `url('${checkDisabled}')`,
  backgroundSize: '7px 7px, 1.9px 1.9px',
});

globalStyle(`${field}:disabled + ${icon}`, {
  backgroundColor: contract.colors.inputBackgroundDisabled,
});
