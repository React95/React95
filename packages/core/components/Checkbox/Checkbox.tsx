import React, { forwardRef } from 'react';

import * as styles from './Checkbox.css';

export type LabelProps = {
  disabled?: boolean;
};
export type CheckboxProps = {
  label?: string;
  children?: string;
  style?: React.CSSProperties;
} & LabelProps &
  React.HTMLProps<HTMLInputElement>;

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  ({ children, style, label, disabled = false, ...rest }, ref) => (
    <label style={style} className={styles.label}>
      <input
        className={styles.field}
        type="checkbox"
        disabled={disabled}
        {...rest}
        ref={ref}
      />
      <span className={styles.icon} />
      <span className={styles.text}>{children || label}</span>
    </label>
  ),
);
