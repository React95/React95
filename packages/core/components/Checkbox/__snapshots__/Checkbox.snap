// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Checkbox /> > Snapshots > should match snapshot with disabled prop 1`] = `
<div>
  <label
    class="Checkbox_label__6vt7xv3"
  >
    <input
      class="Checkbox_field__6vt7xv2"
      disabled=""
      type="checkbox"
    />
    <span
      class="Checkbox_icon__6vt7xv0"
    />
    <span
      class="Checkbox_text__6vt7xv1"
    />
  </label>
</div>
`;

exports[`<Checkbox /> > Snapshots > should match snapshot without props 1`] = `
<div>
  <label
    class="Checkbox_label__6vt7xv3"
  >
    <input
      class="Checkbox_field__6vt7xv2"
      type="checkbox"
    />
    <span
      class="Checkbox_icon__6vt7xv0"
    />
    <span
      class="Checkbox_text__6vt7xv1"
    />
  </label>
</div>
`;
